trie/zk/merkle_tree.go

```diff
@@ -0,0 +1,479 @@
+}
+
+func (t *MerkleTree) TryDelete(key []byte) error {
+	k, err := zkt.ToSecureKey(key)
+	if err != nil {
+		return err
+	}
+	return t.TryDeleteByNodeKey(zkt.NewHashFromBigInt(k))
+}
+
+// TryDeleteByNodeKey removes the specified Key from the MerkleTree and updates the path
+// from the deleted key to the Root with the new values. This method removes
+// the key from the MerkleTree, but does not remove the old nodes from the
+// key-value database; this means that if the tree is accessed by an old Root
+// where the key was not deleted yet, the key will still exist. If is desired
+// to remove the key-values from the database that are not under the current
+// Root, an option could be to dump all the leafs (using mt.DumpLeafs) and
+// import them in a new MerkleTree in a new database (using
+// mt.ImportDumpedLeafs), but this will lose all the Root history of the MerkleTree
+func (t *MerkleTree) TryDeleteByNodeKey(nodeKey *zkt.Hash) error {
+	if !zkt.CheckBigIntInField(nodeKey.BigInt()) { // verify that k is valid and fit inside the Finite Field.
+		return trie.ErrInvalidField
+	}
+	node, path, pathNodes := t.rootNode, t.newTreePath(nodeKey), *new([]*ParentNode)
+	for i := 0; i < t.maxLevels; i++ {
+		switch n := node.(type) {
+		case *ParentNode:
+			pathNodes = append(pathNodes, n)
+			node = t.getChild(n, path.Get(i))
+			continue
+		case *LeafNode:
+			if bytes.Equal(nodeKey[:], n.KeyHash[:]) {
+				t.rmAndUpload(path, pathNodes)
+				return nil
+			}
+			return trie.ErrKeyNotFound
+		case *EmptyNode:
+			return trie.ErrKeyNotFound
+		case *HashNode:
+			return trie.ErrKeyNotFound
+		default:
+			return trie.ErrInvalidNodeFound
+		}
+	}
+	return trie.ErrKeyNotFound
+}
+
+// rmAndUpload removes the key, and goes up until the root updating all the nodes with the new values.
+func (t *MerkleTree) rmAndUpload(path TreePath, pathNodes []*ParentNode) {
+	switch len(pathNodes) {
+	case 0: // root node does not parent node.
+		t.removeNodeHash(t.rootNode)
+		t.rootNode = EmptyNodeValue
+	case 1:
+		t.removeNodeHash(t.rootNode)
+		t.rootNode = t.getChild(pathNodes[0], path.GetOther(0))
+	default:
+		lastSibling := t.getChild(pathNodes[len(pathNodes)-1], path.GetOther(len(pathNodes)-1))
+		defer func() {
+			if t.rootNode != lastSibling {
+				for _, node := range pathNodes {
+					t.removeNodeHash(node)
+				}
+			}
+		}()
+		if _, ok := lastSibling.(*ParentNode); ok {
+			pathNodes[len(pathNodes)-1].SetChild(path.Get(len(pathNodes)-1), EmptyNodeValue)
+			return
+		}
+		for i := len(pathNodes) - 2; i >= 0; i-- {
+			t.removeNodeHash(pathNodes[i])
+			sibling := t.getChild(pathNodes[i], path.GetOther(i))
+			if _, ok := sibling.(*EmptyNode); ok {
+				continue
+			}
+			pathNodes[i].SetChild(path.Get(i), lastSibling)
+			pathNodes = pathNodes[:i]
+			return
+		}
+		// if all sibling is zero, stop and store the sibling of the deleted leaf as root
+		t.rootNode = lastSibling
+	}
+}
+
+func (t *MerkleTree) removeNodeHash(node TreeNode) {
```

Comment:
When does it remove the data from database?

Reply:
First of all, zktrie, which is the reference for the implementation, does not perform a delete on the storage area when you call the try delete function.
So I can't give you a precise answer as to when it does perform a delete.
My best guess is that it deletes when pruning, but I'd have to spend more time checking how it does so in storage, since the node's key doesn't have a prefix, so it can't be iterated over.